store/tikv/lock_resolver.go

```diff
@@ -578,10 +590,147 @@ func (lr *LockResolver) getTxnStatus(bo *Backoffer, txnID uint64, primary []byte
+	tikvLockResolverCountWithResolveAsync.Inc()
+
+	regions, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, status.secondaries, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	commitTs := l.MinCommitTS
+	commitTsMutex := sync.Mutex{}
+	locks := []*Lock{l}
+	missingLock := false
+	errChan := make(chan error, len(regions))
+
+	for regionID, keys := range regions {
+		curRegionID := regionID
+		curKeys := keys
+
+		go func() {
+			for {
+				checkReq := &kvrpcpb.CheckSecondaryLocksRequest{
+					Keys:         curKeys,
+					StartVersion: l.TxnID,
+				}
+				req := tikvrpc.NewRequest(tikvrpc.CmdCheckSecondaryLocks, checkReq)
+				tikvLockResolverCountWithQueryCheckSecondaryLocks.Inc()
+				resp, err := lr.store.SendReq(bo, req, curRegionID, readTimeoutShort)
+				if err != nil {
+					errChan <- errors.Trace(err)
+					return
+				}
+				regionErr, err := resp.GetRegionError()
+				if err != nil {
+					errChan <- errors.Trace(err)
+					return
+				}
+				if regionErr != nil {
+					err = bo.Backoff(BoRegionMiss, errors.New(regionErr.String()))
+					if err != nil {
+						errChan <- errors.Trace(err)
+						return
+					}
+					continue
+				}
+				if resp.Resp == nil {
+					errChan <- errors.Trace(ErrBodyMissing)
+					return
+				}
+				checkResp := resp.Resp.(*kvrpcpb.CheckSecondaryLocksResponse)
+
+				// Check each lock.
+				for _, lockInfo := range checkResp.Locks {
+					if lockInfo == nil {
+						// A lock is missing - the transaction must either be rolled back or committed.
+						commitTsMutex.Lock()
+						// I recommend to add a check if the resp.CommitTs equals to commitTs when missingLock is already true. This may help us to find bugs
+						if !missingLock {
+							commitTs = checkResp.CommitTs
+						}
+						missingLock = true
+						commitTsMutex.Unlock()
+
+						if commitTs != checkResp.CommitTs {
+							errChan <- errors.Errorf("commit TS mismatch in async commit recovery: %v and %v", commitTs, checkResp.CommitTs)
+						} else {
+							errChan <- nil
+						}
+						return
+					}
+
+					if lockInfo.LockVersion != l.TxnID {
+						err = errors.Errorf("unexpected timestamp, expected: %v, found: %v", l.TxnID, lockInfo.LockVersion)
+						logutil.BgLogger().Error("resolveLockAsync error", zap.Error(err))
+						errChan <- err
+						return
+					}
+
+					commitTsMutex.Lock()
+					if !missingLock && lockInfo.MinCommitTs > commitTs {
+						commitTs = lockInfo.MinCommitTs
+					}
+					locks = append(locks, NewLock(lockInfo))
+					commitTsMutex.Unlock()
+				}
+				break
+			}
+			errChan <- nil
+		}()
+	}
+
+	var errs []string
+	for range regions {
+		err1 := <-errChan
+		if err1 != nil {
+			errs = append(errs, err1.Error())
+		}
+	}
+
+	if len(errs) > 0 {
+		return errors.Errorf("async commit recovery (sending CheckSecondaryLocks) finished with errors: %v", errs)
+	}
+
+	status.commitTS = commitTs
```

Comment:
I think we need to mark a `TODO` here for the schema lease check. The crash may happen between prewrite and schema check.